evm/src/executor/inspector/fuzzer.rs

```diff
@@ -0,0 +1,107 @@
+use crate::{
+    fuzz::{invariant::RandomCallGenerator, strategies::EvmFuzzState},
+    utils,
+};
+use bytes::Bytes;
+use ethers::prelude::H160;
+use revm::{db::Database, CallInputs, CallScheme, EVMData, Gas, Inspector, Interpreter, Return};
+
+/// An inspector that can fuzz and collect data for that effect.
+#[derive(Clone, Debug)]
+pub struct Fuzzer {
+    pub generator: RandomCallGenerator,
+    pub fuzz_state: EvmFuzzState,
+    pub collect: bool,
+}
+
+impl<DB> Inspector<DB> for Fuzzer
+where
+    DB: Database,
+{
+    fn call(
+        &mut self,
+        data: &mut EVMData<'_, DB>,
+        call: &mut CallInputs,
+        _: bool,
+    ) -> (Return, Gas, Bytes) {
+        if data.env.tx.caller != call.context.caller {
+            self.reentrancy(call);
+        }
+
+        self.collect = true;
+
+        (Return::Continue, Gas::new(call.gas_limit), Bytes::new())
+    }
+
+    fn step(
+        &mut self,
+        interpreter: &mut Interpreter,
+        _: &mut EVMData<'_, DB>,
+        _is_static: bool,
+    ) -> Return {
+        // We only collect before and after calls
+        if self.collect {
+            self.collect_data(interpreter);
+            self.collect = false;
+        }
+        Return::Continue
+    }
+
+    fn call_end(
+        &mut self,
+        _: &mut EVMData<'_, DB>,
+        _: &CallInputs,
+        remaining_gas: Gas,
+        status: Return,
+        retdata: Bytes,
+        _: bool,
+    ) -> (Return, Gas, Bytes) {
+        self.generator.used = false;
+        self.collect = true;
+
+        (status, remaining_gas, retdata)
+    }
+}
+
+impl Fuzzer {
+    /// Collects `state` and `memory` values into the fuzz dictionary.
+    fn collect_data(&mut self, interpreter: &mut Interpreter) {
+        let mut state = self.fuzz_state.write();
+
+        for slot in interpreter.stack().data() {
+            state.insert(utils::u256_to_h256_be(*slot).into());
+        }
+
+        for index in 0..interpreter.memory.len() / 32 {
+            let mut slot = [0u8; 32];
+            slot.clone_from_slice(interpreter.memory.get_slice(index * 32, 32));
+
+            state.insert(slot);
+        }
+    }
```

Comment:
> are these values only used in the invariant executor, or also in the normal fuzz executor?

only on the invariant executor for now

Reply:
> do we have any ideas on what the memory looks like if we copy everything?

On the `InvariantReentrancy` test we're getting like 56 * 32 bytes. Feels acceptable, but I don't know how it would look like on a bigger project...

At the same time, i was running some tests, `stack` values that we're collecting seem fine, but the `memory` ones don't look so good.
example
```
[evm/src/executor/inspector/fuzzer.rs:108]  "mem value added: 667f9d700000000000000000000000007109709ecfa91a80626ff3989d68f67f" 
```